RFC Merging Process (#4124)
Scille/parsec-cloud: docs/rfcs/1002-merging-process.md

```markdown
<!-- You can erase any parts of this template not applicable to your pull request -->

## Describe your changes

<!-- Why this PR exist ? what is its goal ?
     Give some context about the changes
     Draw attention to the details that should not be overlooked -->

## Checklist

Before you submit this pull request, please make sure to:

- [ ] Keep changes in the pull request as small as possible
   <!-- Move unrelated changes to a new pull request -->
- [ ] Ensure the commit history is sanitized
   <!-- Commits should have a meaningful title and contain a coherent set of changes
        Squash commits that are only relevant to the branch context -->
- [ ] Give a meaningful title to your PR
- [ ] Describe your changes
   <!-- Give some context about the changes
        Draw attention to the details that should not be overlooked -->
- [ ] Link any related issue in the description
   <!-- You can add `closes #<IssueID>` to close the issue when the PR is merged -->
- [ ] Link any dependent pull request in the description

If this PR adds a new feature or fixes a bug:

- [ ] I have added or updated relevant tests

> 

Author: Note: the package will likely updated by `Dependabot`

> 

Author: Otherwise, not everyone can test the change on all platform.

I think we should also add a comment 

```markdown
<!-- If you don't have a specific platform, seek help of someone that have it. -->
```

> 

Reviewer: :+1: for the suggestions

- [ ] I have updated the user documentation
```

## Reviewing a pull request

Before a pull request can be merged, it need to be reviewed to meet our quality
standards.

The review consists in two steps:

- [The automated review](#the-automated-review)
- [The peer review](#the-peer-review)

Also, there are some extra steps for [specific pull requests reviews](#specific-pull-request-reviews).

### The automated review

The goal is to automate the systematic and tedious checks that can be easily
forgotten by a peer.

This step is responsible for:

- Ensuring codebase compiles (or equivalent)
- Executing tests and verifying their result (cargo test, pytest, ...)
- Verifying coding style using the configured linters (rustfmt, clippy, ruff,
  eslint, ...)
- Checking spell issues (cspell)
- Ensuring the app can still be packaged with the proposed change
- Ensuring commit signatures are valid (gpg signed commit)
- Verifying simple checks from the [Pull-request checklist](#the-pull-request-checklist)
- Verifying simple check on commit titles

### The peer review

The goal is not only to ensure quality standards missed in the automated review,

> but also to encourage **team collaboration** and **knowledge sharing**.

Reviewer 2: and application security (A single developer cannot include backdoors without being seen).

Author: ```suggestion
but also to encourage **team collaboration**, **application security** and **knowledge sharing**.
```

> You may need more than one approving review from different code owners if your change modify different code base (i.e. If you modify both rust & python for example).

As a reviewer, your approval involves some extra responsibilities since it is a
pledge of trust to merge the proposed changes in the master branch. Remember to:

- Read and *understand* the proposed changes
- Ask questions (what is the purpose of this ? can you explain me this ?)
- Take the opportunity to provide valuable feedback
- Suggests improvements like a simpler or more effective implementation, code
  refactoring, code comments, fix typos, etc.
- Ensure the changes are well tested
- If necessary, pull the code locally and perform some extra testing

Also:

- Ensure changes are coherent and related to the linked issue or the pull
  request description
  - If necessary, propose to split the changes into smaller pull requests
- Ensure the author followed the [Pull-request checklist](#the-pull-request-checklist)
  (if applicable: check news fragment, user documentation, etc.)

> - Ensure that [commit history is sanitized](#how-to-keep-a-sanitized-commit-history))

Author: ```suggestion
- Ensure that [commit history is sanitized](#how-to-keep-a-sanitized-commit-history)
```

```markdown
<!-- You can erase any parts of this template not applicable to your pull request -->

## Describe your changes

<!-- Why this PR exist ? what is its goal ?
     Give some context about the changes
     Draw attention to the details that should not be overlooked -->

## Checklist

Before you submit this pull request, please make sure to:

- [ ] Keep changes in the pull request as small as possible
   <!-- Move unrelated changes to a new pull request -->
- [ ] Ensure the commit history is sanitized
   <!-- Commits should have a meaningful title and contain a coherent set of changes
        Squash commits that are only relevant to the branch context -->
- [ ] Give a meaningful title to your PR
- [ ] Describe your changes
   <!-- Give some context about the changes
        Draw attention to the details that should not be overlooked -->
- [ ] Link any related issue in the description
   <!-- You can add `closes #<IssueID>` to close the issue when the PR is merged -->
- [ ] Link any dependent pull request in the description

If this PR adds a new feature or fixes a bug:

- [ ] I have added or updated relevant tests

> If this PR modify installation or update packages:

Reviewer: ```suggestion
If this PR modifies installation or updates packages:
```

Author: Done

| Gui          | The PR is related to the GUI                   |
| Test         | The PR is related to the test codebase         |
| **PER CODE** |                                                |
| CI           | The PR affect the CI                           |
| Rust         | The PR affect the Rust code                    |
| Python       | The PR affect the Python code                  |
| Js           | The PR affect the Javascript / Typescript code |
| Java         | The PR affect the Java / Kotlin / Android code |
| **PER OS**   |                                                |
| Linux        | The PR is specific to Linux                    |
| MacOS        | The PR is specific to MacOS                    |
| Windows      | The PR is specific to Windows                  |

Hints:

- If the pull request is related to more than one category in `PER CODE`, this
  may suggests that it needs to be splitted into smaller pull requests.
  > This would not be enforced because sometime it is required to change
  > multiple parts of the code at the same time.

- If the pull request is related to more than one category in `PER OS`, this

>   will certainly indicate that the change is not OS-specific so spliting the }

Author: ```suggestion
  will certainly indicate that the change is not OS-specific so spliting the
```

standards.

The review consists in two steps:

- [The automated review](#the-automated-review)
- [The peer review](#the-peer-review)

Also, there are some extra steps for [specific pull requests reviews](#specific-pull-request-reviews).

### The automated review

The goal is to automate the systematic and tedious checks that can be easily
forgotten by a peer.

This step is responsible for:

- Ensuring codebase compiles (or equivalent)
- Executing tests and verifying their result (cargo test, pytest, ...)
- Verifying coding style using the configured linters (rustfmt, clippy, ruff,
  eslint, ...)
- Checking spell issues (cspell)
- Ensuring the app can still be packaged with the proposed change
- Ensuring commit signatures are valid (gpg signed commit)
- Verifying simple checks from the [Pull-request checklist](#the-pull-request-checklist)
- Verifying simple check on commit titles

### The peer review

The goal is not only to ensure quality standards missed in the automated review,
but also to encourage **team collaboration** and **knowledge sharing**.

Ideally, the review should be done by the [code owner]((#the-code-owner)) of the

> impacted code parts.

Author: @mmmarcos 

Note: If we enable the code owner system, It would be a requirement ;)

Author: Good proposal, I would like to add that anyone is free to do a review, but it's the code owner one that will allow to merge

- Verifying simple check on commit titles

- The proposed change can compile (or equivalent).
- Execute the tests and verify their result (cargo test, pytest, ...).
- Verify the coding style by using the configured linters (rustfmt, clippy, ruff, eslint, ...).
- Verify typo
- Check if we're able to package the app with the proposed change.
- Basic check that we've followed the [Pull-request checklist](#the-pull-request-checklist).
- Check the commit signature (gpg signed commit) are valid.
- Basic check on commit title.
### The peer review

### The manual or human review process
The goal is not only to ensure quality standards missed in the automated review,
but also to encourage **team collaboration** and **knowledge sharing**.

This is the part where one or more person are involved.
This role of this part is:
Ideally, the review should be done by the [code owner]((#the-code-owner)) of the
impacted code parts.

- Check the author of the PR followed the [Pull-request checklist](#the-pull-request-checklist).

> - Check the news fragment content.

Author: @mmmarcos You removed the indication to check the news fragment, why ?

Reviewer: If I am not mistaken I just moved that to line 246/247 (because the news fragment check is part of the checklist)

Author: Indeed

You could also squash every commits in your branch into a single one.
But you must carefully write the commit title and commit body for them to be meaningfull (they will be most likely the PR title/description)

#### The code owner

A *code owner* is a person or a team assigned to be responsible for a specific
component or code parts, usually because it has a deep knownledge about it.

> For example the Python dev are the owner of the python code.

### Specific pull request reviews

This section describes extra steps or things to pay attention to for specific
types of pull requests.

#### Pull request: fix typos

We want the codebase to not be cluttered with typos (syntax, grammar and others).

> Pull requests that those issues are welcome!

Author: ```suggestion
Pull requests that solves those issues are welcome!
```

in mind that this step could prevent (to a certain limit) a potential
supply-chain attack where a fake update is released with malicious code.

## Merging a pull request

After the pull request has been reviewed and accepted, the author is allowed to
merge it into the main branch.

The merge should be done in a specific way:

- The commits should appear on top of the git history (see [Why do we want a linear git history](#why-do-we-want-a-linear-git-history)).
- It should be tested one last time if changes are compatible with the main
  branch.

> This is done by the [merge queue](#what-is-a-merge-queue) that will rebase the
> commits onto the main branch.

Author: ```suggestion
    This is done by the [merge queue](#what-is-a-merge-queue) that will rebase the
    commits onto the main branch.
```

I think the paragraph should be a part of the last element `It should be tested one last time ...`

### Why do we want a linear git history

> Why want a linear git history because it allows to:

Author: ```suggestion
Why want a linear git history ? because it allows to:
```

- Answer the question: *Did **feature A** get introduced before of after **bugfix B** ?*
- Simpler git history graph (everything is in-line).
- `git bisect` work best with a linear history.
- Remove unecessary commit `Merge branch X into Y` that clutter the history.

How to achieve a linear git history?

Git provide multiple methods for that

- `git fast forward` (`git merge --ff-only`):

  When commits are already ahead of the base branch, it will only update the
  branch pointer.

- `git squash` (`git merge --squash`):

  Apply the change onto the destination branch but wont make a commit, this
  allows to create a single commit on top of the current branch.

- `git rebase`

  Will replay commits that are not already on the `upstream` branch on top of
  the `upstream` branch's commit.

### What is a merge queue

A merge queue is a system that will queue pull requests to be merge in a queue :p

A merge queue typically works by creating a temporary branch that include:

- The latest change from the base branch.
- The change introduced by the PRs ahead of the queue.
- The change added by your PR.

> This ensures that we merge pull requests that work with the latest state of the
> base branch.

Author: ```suggestion
This ensures that we merge pull requests that work with the latest state of the
base branch and the previous to be merge PR.
```
